Nested annotation doesn't work with inner class

**Which version of Kotest are you using**

5.2.3

Hi, I'm trying to migrate test framework from Junit5 to kotest. 

The test codes in my project are written in with DCI pattern using `Nested` annotation. To ensure compatibility, I simply added AnnotationSpec on my test classes, including inner class but it goes into failure with following messages

`java.lang.IllegalArgumentException: object is not an instance of declaring class`

I assume that this is because 'inner classes' are not instantiated.

I know there are other specs(styles) which are proper in DCI pattern, but I think this should be fixed if 'annotation spec' is originally intended for easy migration from Junit. thanks

Basically the bug is here 

https://github.com/kotest/kotest/blob/4af10c624be33d39b88d7241765fdffb08758999/kotest-framework/kotest-framework-api/src/jvmMain/kotlin/io/kotest/core/spec/style/AnnotationSpec.kt#L139

Here's what I think is going on

1. `AnnotationSpec` finds all top-level `@Test`s no problem
2. It then finds all nested classes 
    https://github.com/kotest/kotest/blob/4af10c624be33d39b88d7241765fdffb08758999/kotest-framework/kotest-framework-api/src/jvmMain/kotlin/io/kotest/core/spec/style/AnnotationSpec.kt#L90
3. It then creates a lazy test-invocation for each `@Test` in each nested class
    https://github.com/kotest/kotest/blob/4af10c624be33d39b88d7241765fdffb08758999/kotest-framework/kotest-framework-api/src/jvmMain/kotlin/io/kotest/core/spec/style/AnnotationSpec.kt#L139
4. But it doesn't 'switch classes' - `this@AnnotationSpec` refers to the top level annotation spec, NOT the nested one. Therefore, when the test is executed, it fails.

```none
Caused by: java.lang.IllegalArgumentException: object is not an instance of declaring class
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at kotlin.reflect.jvm.internal.calls.CallerImpl$Method.callMethod(CallerImpl.kt:97)
	at kotlin.reflect.jvm.internal.calls.CallerImpl$Method$Instance.call(CallerImpl.kt:113)
	at kotlin.reflect.jvm.internal.KCallableImpl.call(KCallableImpl.kt:108)
	at kotlin.reflect.full.KCallables.callSuspend(KCallables.kt:54)
	at io.kotest.core.spec.style.AnnotationSpec$callNotExpectingException$1.invokeSuspend(AnnotationSpec.kt:132)
	at io.kotest.core.spec.style.AnnotationSpec$callNotExpectingException$1.invoke(AnnotationSpec.kt)
	at io.kotest.core.spec.style.AnnotationSpec$callNotExpectingException$1.invoke(AnnotationSpec.kt)
	at io.kotest.engine.test.TestCaseExecutor$execute$innerExecute$1.invokeSuspend(TestCaseExecutor.kt:81)
	at io.kotest.engine.test.TestCaseExecutor$execute$innerExecute$1.invoke(TestCaseExecutor.kt)
	at io.kotest.engine.test.TestCaseExecutor$execute$innerExecute$1.invoke(TestCaseExecutor.kt)
	at io.kotest.engine.test.interceptors.CoroutineDebugProbeInterceptor.intercept(CoroutineDebugProbeInterceptor.kt:29)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invokeSuspend(TestCaseExecutor.kt:90)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invoke(TestCaseExecutor.kt)
	at io.kotest.engine.test.TestCaseExecutor$execute$2$1.invoke(TestCaseExecutor.kt)
	at io.kotest.engine.test.interceptors.InvocationTimeoutInterceptor$intercept$3.invokeSuspend(InvocationTimeoutInterceptor.kt:43)
...
```

I'm not sure how to fix this. Maybe add a `KClass` parameter to `callExpectingException(...)`? Or refactor `AnnotationSpec` so it inherits from something like `ContainerScope`, and `@Nested` will register a container? 

I've made a PR with a failing test, and a couple of tidy-ups. I'm not sure how to fix it, so I'll make the PR in case someone else finds my progress so far useful.

--- a/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/core/spec/style/AnnotationSpec.kt
+++ b/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/core/spec/style/AnnotationSpec.kt
@@ -18,6 +18,13 @@ typealias Test = AnnotationSpec.Test
 
 abstract class AnnotationSpec : Spec() {
 
+   // we need to track instances because we have to create instances of nested classes
+   private val instances = mutableMapOf<KClass<*>, AnnotationSpec>()
+
+   init {
+      instances.put(this::class, this)
+   }
+
    override suspend fun beforeSpec(spec: Spec) {
       executeBeforeSpecFunctions()
    }
@@ -50,37 +57,30 @@ abstract class AnnotationSpec : Spec() {
       if (it.isSuspend) it.callSuspend(this) else it.call(this)
    }
 
-   private fun KFunction<*>.toIgnoredRootTest(klass: KClass<*>): RootTest {
-      return deriveRootTest(true, klass)
+   private fun KFunction<*>.toIgnoredRootTest(spec: AnnotationSpec): RootTest {
+      return deriveRootTest(true, spec)
    }
 
-   private fun KFunction<*>.toEnabledRootTest(klass: KClass<*>): RootTest {
-      return deriveRootTest(false, klass)
+   private fun KFunction<*>.toEnabledRootTest(spec: AnnotationSpec): RootTest {
+      return deriveRootTest(false, spec)
    }
 
-   private fun KFunction<*>.deriveRootTest(disabled: Boolean, klass: KClass<*>): RootTest {
-      return if (this.isExpectingException()) {
+   private fun KFunction<*>.deriveRootTest(disabled: Boolean, spec: Spec): RootTest {
+      val test = if (this.isExpectingException()) {
          val expected = this.getExpectedException()
-         RootTest(
-            name = TestNameBuilder.builder(name).build(),
-            test = callExpectingException(expected, klass),
-            source = sourceRef(),
-            type = TestType.Test,
-            config = null,
-            disabled = disabled,
-            factoryId = null,
-         )
+         createTestFnExceptingException(expected, spec)
       } else {
-         RootTest(
-            name = TestNameBuilder.builder(name).build(),
-            test = callNotExpectingException(klass),
-            source = sourceRef(),
-            type = TestType.Test,
-            config = null,
-            disabled = disabled,
-            factoryId = null,
-         )
+         createTestFunctionNotExpectingException(spec)
       }
+      return RootTest(
+         name = TestNameBuilder.builder(name).build(),
+         test = test,
+         source = sourceRef(),
+         type = TestType.Test,
+         config = null,
+         disabled = disabled,
+         factoryId = null,
+      )
    }
 
    override fun rootTests(): List<RootTest> {
@@ -98,12 +98,13 @@ abstract class AnnotationSpec : Spec() {
    }
 
    private fun KClass<*>.findRootTests(): List<RootTest> {
+      val spec = instances.getOrPut(this) { this.java.constructors.first().newInstance() as AnnotationSpec }
       return findTestFunctions().map { f ->
          f.isAccessible = true
          if (f.isIgnoredTest()) {
-            f.toIgnoredRootTest(this)
+            f.toIgnoredRootTest(spec)
          } else {
-            f.toEnabledRootTest(this)
+            f.toEnabledRootTest(spec)
          }
       }
    }
@@ -114,10 +115,13 @@ abstract class AnnotationSpec : Spec() {
          .flatMap { it.findRootTests() }
    }
 
-   private fun KFunction<*>.callExpectingException(expected: KClass<out Throwable>, klass: KClass<*>): suspend TestScope.() -> Unit {
+   private fun KFunction<*>.createTestFnExceptingException(
+      expected: KClass<out Throwable>,
+      spec: Spec,
+   ): suspend TestScope.() -> Unit {
       return {
          val thrown = try {
-            callSuspend(this@AnnotationSpec)
+            callSuspend(spec)
             null
          } catch (t: Throwable) {
             t.unwrapIfReflectionCall()
@@ -127,10 +131,12 @@ abstract class AnnotationSpec : Spec() {
       }
    }
 
-   private fun KFunction<*>.callNotExpectingException(klass: KClass<*>): suspend TestScope.() -> Unit {
+   private fun KFunction<*>.createTestFunctionNotExpectingException(
+      spec: Spec
+   ): suspend TestScope.() -> Unit {
       return {
          try {
-            callSuspend(klass)
+            callSuspend(spec)
          } catch (t: Throwable) {
             throw t.unwrapIfReflectionCall()
          }

--- a/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/core/spec/style/AnnotationSpec.kt
+++ b/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/core/spec/style/AnnotationSpec.kt
@@ -50,20 +50,20 @@ abstract class AnnotationSpec : Spec() {
       if (it.isSuspend) it.callSuspend(this) else it.call(this)
    }
 
-   private fun KFunction<*>.toIgnoredRootTest(): RootTest {
-      return deriveRootTest(true)
+   private fun KFunction<*>.toIgnoredRootTest(klass: KClass<*>): RootTest {
+      return deriveRootTest(true, klass)
    }
 
-   private fun KFunction<*>.toEnabledRootTest(): RootTest {
-      return deriveRootTest(false)
+   private fun KFunction<*>.toEnabledRootTest(klass: KClass<*>): RootTest {
+      return deriveRootTest(false, klass)
    }
 
-   private fun KFunction<*>.deriveRootTest(disabled: Boolean): RootTest {
+   private fun KFunction<*>.deriveRootTest(disabled: Boolean, klass: KClass<*>): RootTest {
       return if (this.isExpectingException()) {
          val expected = this.getExpectedException()
          RootTest(
             name = TestNameBuilder.builder(name).build(),
-            test = callExpectingException(expected),
+            test = callExpectingException(expected, klass),
             source = sourceRef(),
             type = TestType.Test,
             config = null,
@@ -73,7 +73,7 @@ abstract class AnnotationSpec : Spec() {
       } else {
          RootTest(
             name = TestNameBuilder.builder(name).build(),
-            test = callNotExpectingException(),
+            test = callNotExpectingException(klass),
             source = sourceRef(),
             type = TestType.Test,
             config = null,
@@ -101,9 +101,9 @@ abstract class AnnotationSpec : Spec() {
       return findTestFunctions().map { f ->
          f.isAccessible = true
          if (f.isIgnoredTest()) {
-            f.toIgnoredRootTest()
+            f.toIgnoredRootTest(this)
          } else {
-            f.toEnabledRootTest()
+            f.toEnabledRootTest(this)
          }
       }
    }
@@ -114,7 +114,7 @@ abstract class AnnotationSpec : Spec() {
          .flatMap { it.findRootTests() }
    }
 
-   private fun KFunction<*>.callExpectingException(expected: KClass<out Throwable>): suspend TestScope.() -> Unit {
+   private fun KFunction<*>.callExpectingException(expected: KClass<out Throwable>, klass: KClass<*>): suspend TestScope.() -> Unit {
       return {
          val thrown = try {
             callSuspend(this@AnnotationSpec)
@@ -127,10 +127,10 @@ abstract class AnnotationSpec : Spec() {
       }
    }
 
-   private fun KFunction<*>.callNotExpectingException(): suspend TestScope.() -> Unit {
+   private fun KFunction<*>.callNotExpectingException(klass: KClass<*>): suspend TestScope.() -> Unit {
       return {
          try {
-            callSuspend(this@AnnotationSpec)
+            callSuspend(klass)
          } catch (t: Throwable) {
             throw t.unwrapIfReflectionCall()
          }
@@ -262,23 +262,23 @@ abstract class AnnotationSpec : Spec() {
 }
 
 internal fun KClass<out AnnotationSpec>.findBeforeTestFunctions() =
-   findFunctionAnnotatedWithAnyOf(AnnotationSpec.BeforeEach::class, AnnotationSpec.Before::class)
+   findMethodsAnnotatedWithAnyOf(AnnotationSpec.BeforeEach::class, AnnotationSpec.Before::class)
 
 internal fun KClass<out AnnotationSpec>.findBeforeSpecFunctions() =
-   findFunctionAnnotatedWithAnyOf(AnnotationSpec.BeforeAll::class, AnnotationSpec.BeforeClass::class)
+   findMethodsAnnotatedWithAnyOf(AnnotationSpec.BeforeAll::class, AnnotationSpec.BeforeClass::class)
 
 internal fun KClass<out AnnotationSpec>.findAfterSpecFunctions() =
-   findFunctionAnnotatedWithAnyOf(AnnotationSpec.AfterAll::class, AnnotationSpec.AfterClass::class)
+   findMethodsAnnotatedWithAnyOf(AnnotationSpec.AfterAll::class, AnnotationSpec.AfterClass::class)
 
 internal fun KClass<out AnnotationSpec>.findAfterTestFunctions() =
-   findFunctionAnnotatedWithAnyOf(AnnotationSpec.AfterEach::class, AnnotationSpec.After::class)
+   findMethodsAnnotatedWithAnyOf(AnnotationSpec.AfterEach::class, AnnotationSpec.After::class)
 
 internal fun KClass<*>.findTestFunctions(): List<KFunction<*>> =
-   findFunctionAnnotatedWithAnyOf(AnnotationSpec.Test::class)
+   findMethodsAnnotatedWithAnyOf(AnnotationSpec.Test::class)
 
 internal fun KFunction<*>.isIgnoredTest() = isFunctionAnnotatedWithAnyOf(AnnotationSpec.Ignore::class)
 
-internal fun KClass<*>.findFunctionAnnotatedWithAnyOf(vararg annotation: KClass<*>): List<KFunction<*>> =
+internal fun KClass<*>.findMethodsAnnotatedWithAnyOf(vararg annotation: KClass<*>): List<KFunction<*>> =
    memberFunctions.filter { it.isFunctionAnnotatedWithAnyOf(*annotation) }
 
 internal fun KFunction<*>.isFunctionAnnotatedWithAnyOf(vararg annotation: KClass<*>) =

--- a/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/spec/annotation/AnnotationSpecNestedTest.kt
+++ b/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/spec/annotation/AnnotationSpecNestedTest.kt
@@ -1,0 +1,36 @@
+package com.sksamuel.kotest.engine.spec.annotation
+
+import io.kotest.core.spec.style.AnnotationSpec
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.engine.TestEngineLauncher
+import io.kotest.engine.listener.CollectingTestEngineListener
+import io.kotest.matchers.maps.shouldHaveSize
+import io.kotest.matchers.shouldBe
+
+class AnnotationSpecNestedTest : FunSpec() {
+   init {
+      test("should detect nested classes") {
+         val listener = CollectingTestEngineListener()
+         TestEngineLauncher(listener).withClasses(AnnotationSpecWithNested::class).launch()
+         listener.tests.shouldHaveSize(2)
+         listener.tests.keys.map { it.name.name }.toSet() shouldBe setOf("foo", "bar")
+         listener.result("foo")!!.isSuccess shouldBe true
+         listener.result("bar")!!.isSuccess shouldBe true
+      }
+   }
+}
+
+private class AnnotationSpecWithNested : AnnotationSpec() {
+   @Test
+   fun foo() {
+      println("foo")
+   }
+
+   @Nested
+   class InsideTest : AnnotationSpec() {
+      @Test
+      fun bar() {
+         println("bar")
+      }
+   }
+}

--- a/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/spec/annotation/AnnotationSpecTest.kt
+++ b/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/spec/annotation/AnnotationSpecTest.kt
@@ -21,12 +21,6 @@ class AnnotationSpecTest : DescribeSpec({
          listener.tests.shouldHaveSize(2)
       }
 
-//      it("should detect nested classes") {
-//         val listener = CollectingTestEngineListener()
-//         TestEngineLauncher(listener).withClasses(AnnotationSpecWithNested::class).launch()
-//         listener.tests.shouldHaveSize(2)
-//      }
-
       it("should support throwing exceptions with @Test(expected=foo)") {
          val listener = CollectingTestEngineListener()
          TestEngineLauncher(listener).withClasses(AnnotationSpecWithExceptions::class).launch()
@@ -82,17 +76,3 @@ private class AnnotationSpecWithExceptions : AnnotationSpec() {
       // Throw nothing
    }
 }
-
-//class AnnotationSpecWithNested : AnnotationSpec() {
-//
-//   @Test
-//   fun foo() {
-//   }
-//
-//   @Nested
-//   class MyNested : AnnotationSpec() {
-//      @Test
-//      fun bar() {
-//      }
-//   }
-//}

